src/adapter/src/coord/peek.rs

```diff
@@ -423,6 +493,102 @@ impl crate::coord::Coordinator {
+
+                    // Re-used state for processing and building rows.
+                    let mut accum = vec![];
+                    let mut datum_vec = DatumVec::new();
+                    let mut row_builder = Row::default();
+                    let arena = RowArena::new();
+
+                    let start = Instant::now();
+                    let mut last_fetch = Duration::ZERO;
+                    while limit_remaining > 0 {
+                        let Some(batch) = cursor.next().await else {
+                            break;
+                        };
+                        last_fetch = start.elapsed();
+                        for ((k, v), _, d) in batch {
+                            let row = k?.0.map_err(|e| e.to_string())?;
+                            let () = v?;
+                            let count: usize = d
+                                .try_into()
+                                .map_err(|_| "Negative count for thing in snapshot".to_owned())?;
+                            let Some(count) = NonZeroUsize::new(count) else {
+                                continue;
+                            };
+                            let mut datum_local = datum_vec.borrow_with(&row);
+                            let eval_result = mfp_plan
+                                .evaluate_into(&mut datum_local, &arena, &mut row_builder)
+                                .map_err(|e| e.to_string())?;
+                            if let Some(row) = eval_result {
+                                accum.push((row, count));
+                                limit_remaining = limit_remaining.saturating_sub(count.get());
+                                if limit_remaining == 0 {
+                                    break;
+                                }
+                            }
+                        }
+                    }
+
+                    let res = finishing.finish(accum, max_result_size)?;
+                    let total_duration = start.elapsed();
+                    metrics
+                        .persist_peek_seconds
+                        .observe(total_duration.as_secs_f64());
+                    info!(
+                        collection =? id,
+                        fetch_duration =? last_fetch,
+                        total_duration =? total_duration,
+                        "persist peek success"
+                    );
+                    Ok(PeekResponseUnary::Rows(res))
+                });
+            let internal_cmd_tx = self.internal_cmd_tx.clone();
+            let ctx_extra = mem::take(ctx_extra);
+            return Ok(ExecuteResponse::SendingRows {
+                future: Box::pin(async move {
+                    let res = match handle.await {
+                        Ok(Ok(res)) => res,
+                        Ok(Err(err)) => PeekResponseUnary::Error(err.to_string()),
+                        Err(_join_error) => PeekResponseUnary::Canceled,
```

Comment:
If the task gets cancelled. (In theory we'd get panic errors here too, but we abort on panic.) This shouldn't happen in the existing code much, since we don't have proper cancellation, but I figured I might as well wire it up. (IIRC it may also happen if the async runtime is being shut down, but I think we don't care that much about what happens in that case.)